Client.py

```diff
@@ -0,0 +1,240 @@
+                        self._parse_header(lines[-1])
+                        self._recv_buff = ""
+                    else:
+                        self._recv_buff = lines[-1]
+
+                    self.logger.debug("Now recv_buff is %s" % self._recv_buff)
+
+        except CustomExceptions.Disconnect:
+            raise
+        except OSError as e:
+            self.logger.error(traceback.format_exc())
+            if e.errno == errno.ENOENT:
+                self._send_buff = util.creat_error(404, 'File Not Found', e)
+                self.request = Request.Request()
+                self.request.full_request = True
+                self._recv_buff = ""
+            else:
+                self._send_buff = util.creat_error(500, 'Internal Error', e)
+                self.request = Request.Request()
+                self.request.full_request = True
+                self._recv_buff = ""
+        except Exception as e:
+            self.logger.error(traceback.format_exc())
+            raise
+
+    def send(self):
+        """ Fill self.send_buff with all the data it needs and then send it
+        """
+        if self.check_finished_request():
+            raise CustomExceptions.FinishedRequest
+        free_space_in_buffer = self._buff_size - len(self._send_buff)
+        if free_space_in_buffer > 0 and self.file is not None:
+            self._send_buff += self.file.read_buff(free_space_in_buffer)
+        if len(self._send_buff) > 0:
+            maybe_sent_status = False
+            if constants.DOUBLE_CRLF in self._send_buff:
+                maybe_sent_status = True
+            self._send_my_buff()
+            if constants.DOUBLE_CRLF in self._send_buff:
+                maybe_sent_status = False
+            if maybe_sent_status:
+                self.request.sent_status = True
+        if self.check_finished_request():
+            raise CustomExceptions.FinishedRequest
+
+    def check_finished_request(self):
+        """Check if we finished the request"""
+        if self.request.get_uri() is None:
+            ans = self.request.sent_status and not self._send_buff
+        else:
+            ans = self.request.sent_status and self.file.check_read_all() and\
+                not self._send_buff
+        return ans
+
+    def _send_my_buff(self):
+        """Send the data in self._send_buff"""
+        self.logger.debug("start sending my buff, send_buff %s" %
+                          self._send_buff)
+        while self._send_buff:
+            try:
+                self._send_buff = self._send_buff[
+                    self._socket.send(self._send_buff):]
+                self.logger.debug("client sent")
+            except socket.error as e:
+                if e.errno not in (errno.EWOULDBLOCK, errno.EAGAIN):
+                    self.logger.error(traceback.format_exc())
+                    raise
+                else:
+                    self.logger.debug(traceback.format_exc())
+                    break
```

Comment:
in both cases you terminate the loop, so probably you can think of a simpler algorithm, hint: you do not need this break